File: Sources/CodeRim/Sources/UsageNormalizer.swift
import Foundation

struct UsageNormalizationState: Equatable, Sendable {
    var cumulativeHighWaterMark: TokenUsage?
    var lastObservedAt: Date?
    var quality: DataQuality

    init(
        cumulativeHighWaterMark: TokenUsage?,
        lastObservedAt: Date? = nil,
        quality: DataQuality
    ) {
        self.cumulativeHighWaterMark = cumulativeHighWaterMark
        self.lastObservedAt = lastObservedAt
        self.quality = quality
    }

    static let empty = UsageNormalizationState(
        cumulativeHighWaterMark: nil,
        lastObservedAt: nil,
        quality: .exact
    )
}

struct UsageNormalizationResult: Equatable, Sendable {
    let delta: TokenUsage?
    let state: UsageNormalizationState
    let diagnostic: String?
}

struct UsageNormalizer: Sendable {
    func normalize(
        _ observation: CodexTokenObservation,
        metadata: SessionMetadata?,
        state: UsageNormalizationState
    ) -> UsageNormalizationResult {
        if let lastObservedAt = state.lastObservedAt,
           observation.occurredAt < lastObservedAt {
            return UsageNormalizationResult(
                delta: nil,
                state: state,
                diagnostic: "out-of-order token snapshot ignored"
            )
        }

        guard let cumulative = observation.cumulativeUsage else {
            return UsageNormalizationResult(
                delta: nil,
                state: UsageNormalizationState(
                    cumulativeHighWaterMark: state.cumulativeHighWaterMark,
                    lastObservedAt: state.lastObservedAt,
                    quality: .partial
                ),
                diagnostic: "missing cumulative token usage"
            )
        }

        guard let previous = state.cumulativeHighWaterMark else {
            guard observation.lastUsage == cumulative else {
                return UsageNormalizationResult(
                    delta: nil,
                    state: UsageNormalizationState(
                        cumulativeHighWaterMark: cumulative,
                        lastObservedAt: observation.occurredAt,
                        quality: .partial
                    ),
                    diagnostic: "initial cumulative baseline is unresolved"
                )
            }

            return UsageNormalizationResult(
                delta: cumulative.isZero ? nil : cumulative,
                state: UsageNormalizationState(
                    cumulativeHighWaterMark: cumulative,
                    lastObservedAt: observation.occurredAt,
                    quality: state.quality
                ),
                diagnostic: nil
            )
        }

        guard cumulative.isComponentWiseAtLeast(previous) else {
            if observation.lastUsage == cumulative {
                return UsageNormalizationResult(
                    delta: cumulative.isZero ? nil : cumulative,
                    state: UsageNormalizationState(
                        cumulativeHighWaterMark: cumulative,
                        lastObservedAt: observation.occurredAt,
                        quality: .partial
                    ),
                    diagnostic: "cumulative counter restarted"
                )
            }

            return UsageNormalizationResult(
                delta: nil,
                state: UsageNormalizationState(
                    cumulativeHighWaterMark: previous.componentWiseMaximum(with: cumulative),
                    lastObservedAt: observation.occurredAt,
                    quality: .partial
                ),
                diagnostic: "cumulative token usage decreased or interleaved"
            )
        }

        let delta = cumulative.subtractingFloorAtZero(previous)
        return UsageNormalizationResult(
            delta: delta.isZero ? nil : delta,
            state: UsageNormalizationState(
                cumulativeHighWaterMark: cumulative,
                lastObservedAt: observation.occurredAt,
                quality: state.quality
            ),
            diagnostic: nil
        )
    }
}

File: Windows/src/CodeRim.Core/Services/UsageNormalizer.cs
using CodeRim.Core.Domain;

namespace CodeRim.Core.Services;

public sealed record UsageNormalizationState(
    TokenUsage? CumulativeHighWaterMark,
    DateTimeOffset? LastObservedAt,
    DataQuality Quality)
{
    public static UsageNormalizationState Empty { get; } = new(null, null, DataQuality.Exact);
}

public sealed record UsageNormalizationResult(
    TokenUsage? Delta,
    UsageNormalizationState State,
    string? Diagnostic);

public static class UsageNormalizer
{
    public static UsageNormalizationResult Normalize(
        TokenObservation observation,
        UsageNormalizationState state)
    {
        ArgumentNullException.ThrowIfNull(observation);
        ArgumentNullException.ThrowIfNull(state);

        if (state.LastObservedAt is not null
            && observation.OccurredAt < state.LastObservedAt.Value)
        {
            return new UsageNormalizationResult(
                null,
                state,
                "out-of-order token snapshot ignored");
        }

        if (observation.CumulativeUsage is not { } cumulative)
        {
            return new UsageNormalizationResult(
                null,
                state with { Quality = DataQuality.Partial },
                "missing cumulative token usage");
        }

        if (state.CumulativeHighWaterMark is not { } previous)
        {
            if (observation.LastUsage != cumulative)
            {
                return new UsageNormalizationResult(
                    null,
                    new UsageNormalizationState(cumulative, observation.OccurredAt, DataQuality.Partial),
                    "initial cumulative baseline is unresolved");
            }

            return new UsageNormalizationResult(
                cumulative.IsZero ? null : cumulative,
                new UsageNormalizationState(cumulative, observation.OccurredAt, state.Quality),
                null);
        }

        if (cumulative.HasCounterDecrease(previous))
        {
            if (observation.LastUsage == cumulative)
            {
                return new UsageNormalizationResult(
                    cumulative.IsZero ? null : cumulative,
                    new UsageNormalizationState(cumulative, observation.OccurredAt, DataQuality.Partial),
                    "cumulative counter restarted");
            }

            return new UsageNormalizationResult(
                null,
                new UsageNormalizationState(
                    previous.ComponentWiseMaximum(cumulative),
                    observation.OccurredAt,
                    DataQuality.Partial),
                "cumulative token usage decreased or interleaved");
        }

        var delta = cumulative.SubtractFloorAtZero(previous);
        return new UsageNormalizationResult(
            delta.IsZero ? null : delta,
            new UsageNormalizationState(cumulative, observation.OccurredAt, state.Quality),
            null);
    }
}
